File: ping/tao-ping-client.cpp
#include <iostream>

#include "PingC.h"

int main(int argc, char** argv) {
    try {
        CORBA::ORB_var orb = CORBA::ORB_init(argc, argv);
        CORBA::Object_var po = orb->string_to_object("corbaname::nameservice/NameService#test/ExamplePing");
        Example::Pingable_var pingable = Example::Pingable::_narrow(po.in());
        pingable->ping();
        orb->destroy();
    } catch (CORBA::Exception const& e) {
        std::cerr << "CORBA exception raised: " << e._name() << ": " << e._info().c_str() << '\n';
    }
    return 0;
}

File: ping/tao_ping_client.nim
{.compile: "PingC.cpp".}

import os

const pingc = "/source/PingC.h"

type
  # CorbaOrbPtr {.final, header: pingc, importcpp: "CORBA::ORB_ptr".} = object
  CorbaOrbVar {.final, header: pingc, importcpp: "CORBA::ORB_var".} = object
  CorbaObjectVar {.final, header:pingc, importcpp: "CORBA::Object_var".} =
    object
  ExamplePingableVar {.final, header:pingc,
                       importcpp: "Example::Pingable_var".} = object
  CorbaTransient {.final, header:pingc, importcpp: "CORBA::TRANSIENT".} = object of Exception

proc corbaOrbInit(argc: cint, argv: cstringArray): CorbaOrbVar
  {.header: pingc, importcpp: "CORBA::ORB_init(@)".}

proc stringToObject(orb: CorbaOrbVar, s: cstring): CorbaObjectVar
  {.header: pingc, importcpp: "#->string_to_object(@)".}

proc destroy(orb: CorbaOrbVar)
  {.header: pingc, importcpp: "#->destroy(@)".}

proc examplePingableNarrow(obj: CorbaObjectVar): ExamplePingableVar
  {.header: pingc, importcpp: "Example::Pingable::_narrow(@)".}

# I don't yet know how to get C++ exceptions to propagate to Nim, so the raises
# pragma is useless.
proc ping(pingable: ExamplePingableVar)
  {.header: pingc, importcpp: "#->ping(@)", raises: [CorbaTransient].}

var params = os.commandLineParams()
var argc: cint = params.len().cint
var argv: cstringArray = allocCStringArray(params)

try:
  var orb = corbaOrbInit(argc, argv)
  var po = orb.stringToObject("corbaname::nameservice/NameService#test/ExamplePing")
  var pingable = examplePingableNarrow(po)
  pingable.ping()
  orb.destroy()
except CorbaTransient:
  echo("Got an exception!")